Support Darwin and AF_VSOCK.

ijc opened the pull request:
Significantly refactor to support vsock and Darwin. Matrix is:

```
        | hvsock | vsock |
--------+--------+-------+
Darwin  |        |   X   |
Linux   |   X    |   X   |
Windows |   X    |       |
```

Linux tries to autodetect AF_VSOCK vs AF_HYPERV but can be overridden with `-m
vsock` or `-m hvsock`.

Darwin assumes Hyperkit as configured by Docker for Mac, but can be overridden
with e.g. `-m hyperkit:/some/path` for standalone hyperkit.

Add a README to explain how to run things. Note that I have only build tested
for Windows and it remains undocumented.

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

Commit:
Support Darwin and AF_VSOCK.

Significantly refactor to support vsock and Darwin. Matrix is:

        | hvsock | vsock |
--------+--------+-------+
Darwin  |        |   X   |
Linux   |   X    |   X   |
Windows |   X    |       |

Linux tries to autodetect AF_VSOCK vs AF_HYPERV but can be overridden with `-m
vsock` or `-m hvsock`.

Darwin assumes Hyperkit as configured by Docker for Mac, but can be overridden
with e.g. `-m hyperkit:/some/path` for standalone hyperkit.

Add a README to explain how to run things. Note that I have only build tested
for Windows and it remains undocumented.

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

ijc25 commented:
Of particular note above:
> Note that I have only build tested for Windows and it remains undocumented.

ijc25 commented:
I meant to mention this in the commit message but forgot: It was easier to add a verbosity option to `hvgoecho` than to faff around refactoring it even further to allow only `hvgostress` to have the option.

Commit:
Use a single Reader for the entire client side

Otherwise the "BYE" message can be cached inside the first and never read
because the code which looks for it constructs a new one.

This solves occasional failure to see the BYE:

    2016/11/29 12:30:47 [00000] TX: 42103 bytes, md5=a16973fb30a5d4b970687522a3935197
    2016/11/29 12:30:47 [00000] RX: 42103 bytes, md5=a16973fb30a5d4b970687522a3935197 (sent=42103)
    2016/11/29 12:30:47 [00000] Failed to receive bye: EOF
    2016/11/29 12:30:47 [00000] From SVR:

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

ijc25 commented:
+ a bonus bugfix, missing BYE message due to reusing the Reader on the client side.

rneugeba commented:
Thanks. Very useful.

I tried testing `hvgostress` on the Windows side:
- Server on Linux/Client on Windows works
- Server on Windows returns:
```
PS C:\Users\rneugeba\Desktop> .\hvgostress.exe -s
Starting server
2016/11/29 15:37:16 Accept(): accept(): Unimplemented
```

rneugeba commented:
Doh, forgot that I didn't implement `accept()` on Windows due to issues with connecting from the VM to the host. Since the `host`->`Linux VM` works on Windows. Merging this.
